numpy/_core/src/multiarray/datetime.c

```diff
@@ -2122,8 +2125,7 @@ NpyDatetime_ConvertPyDateTimeToDatetimeStruct(
     }
 
     /* Get the hour */
-    tmp = PyObject_GetAttrString(obj, "hour");
-    if (tmp == NULL) {
+    if (PyObject_GetOptionalAttr(obj, npy_interned_str.hour, &tmp) != 1) {
```

Comment:
This does the wrong thing if the return value is `-1`; it returns `-1` without an exception set.

Either you need to put this back to `PyObject_GetAttr`, or you  need to merge this with the `PyObject_HasAttrString` above as you did for `year`/`month`/`day`.